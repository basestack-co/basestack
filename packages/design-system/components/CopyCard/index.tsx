import { memo, useState } from "react";
import type { LayoutProps, SpaceProps } from "styled-system";
import IconButton from "../IconButton";
// Components
import Text from "../Text";
import { Tooltip, TooltipContent, TooltipTrigger } from "../Tooltip";
import {
  Container,
  DescriptionContainer,
  IconButtonContainer,
  TitleContainer,
} from "./styles";

export interface CopyCardProps extends LayoutProps, SpaceProps {
  title?: string;
  description: string;
  tooltip: {
    defaultText: string;
    successText: string;
  };
}

const CopyCard = ({ title, description, tooltip, ...props }: CopyCardProps) => {
  const [showTooltipSuccess, setShowTooltipSuccess] = useState(false);

  return (
    <Container {...props}>
      <TitleContainer>
        <Text>{title}</Text>
      </TitleContainer>
      <DescriptionContainer>
        <Text lineTruncate>{description}</Text>
      </DescriptionContainer>
      <IconButtonContainer>
        <Tooltip placement="top">
          <TooltipTrigger onMouseLeave={() => setShowTooltipSuccess(false)}>
            <IconButton
              variant="neutral"
              icon="content_copy"
              onClick={async () => {
                await navigator.clipboard.writeText(description);
                setShowTooltipSuccess(true);
              }}
            />
          </TooltipTrigger>
          <TooltipContent>
            {showTooltipSuccess ? tooltip.successText : tooltip.defaultText}
          </TooltipContent>
        </Tooltip>
      </IconButtonContainer>
    </Container>
  );
};

export default memo(CopyCard);
