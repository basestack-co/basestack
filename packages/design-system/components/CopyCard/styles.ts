import { rem } from "polished";
import styled, { css } from "styled-components";
import { compose, layout, space } from "styled-system";

export const Container = styled.div.withConfig({
  shouldForwardProp: (prop) => prop !== "maxWidth",
})`
  height: ${rem("44px")};
  display: flex;
  background-color: ${({ theme }) => theme.copyCard.backgroundColor};
  border-radius: 4px;
  overflow: hidden;
  ${compose(layout, space)};
`;

const containerSharedStyles = css`
  display: flex;
  align-items: center;
`;

export const TitleContainer = styled.div`
  ${containerSharedStyles};
  background-color: ${({ theme }) => theme.copyCard.title.backgroundColor};
  padding: 0 ${({ theme }) => theme.spacing.s4};
  flex-shrink: 0;
`;

export const DescriptionContainer = styled.div`
  ${containerSharedStyles};
  padding: 0 ${({ theme }) => theme.spacing.s4};
  overflow: hidden;
`;

export const IconButtonContainer = styled.div`
  ${containerSharedStyles};
  padding-right: ${({ theme }) => theme.spacing.s2};
  margin-left: auto;
`;
