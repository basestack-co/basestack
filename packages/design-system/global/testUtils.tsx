import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
