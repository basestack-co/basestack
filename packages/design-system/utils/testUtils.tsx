import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
