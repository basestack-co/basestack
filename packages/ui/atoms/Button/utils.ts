import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
