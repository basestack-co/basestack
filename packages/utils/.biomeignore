# Node dependencies
node_modules

# Next.js build output
.next
out

# Environment files
.env
.env.*
!.env.example

# Logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
pnpm-debug.log*

# OS / editor files
.DS_Store
Thumbs.db

# Package manager cache
.pnp
.pnp.js
.pnp.cjs

# Test coverage
coverage

# Build artifacts
dist
build

# Temporary files
.tmp
temp