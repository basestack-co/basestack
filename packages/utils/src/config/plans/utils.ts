export const getAppMode = (mode: string) => {
  return (
    ["development", "production", "local"].includes(mode) ? mode : "production"
  ) as "local" | "development" | "production";
};
