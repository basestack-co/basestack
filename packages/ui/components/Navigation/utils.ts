// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
