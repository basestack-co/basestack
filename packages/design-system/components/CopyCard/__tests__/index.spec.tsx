import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import CopyCard from "..";

describe("CopyCard tests", () => {
  afterEach(cleanup);

  test("should render CopyCard correctly", () => {
    const { asFragment } = renderWithTheme(
      <CopyCard
        title="title"
        description="description"
        tooltip={{
          defaultText: "defaultText",
          successText: "successText",
        }}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
