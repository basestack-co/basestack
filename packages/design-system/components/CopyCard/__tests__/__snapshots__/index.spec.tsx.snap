// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CopyCard tests should render CopyCard correctly 1`] = `
<DocumentFragment>
  .c8 {
  height: 1.5rem;
  width: 1.5rem;
  flex-shrink: 0;
}

.c9 {
  color: #000000;
  font-size: 1.5rem;
}

.c7 {
  border: none;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out,border-color 0.1s ease-in-out;
  background-color: transparent;
  color: #000000;
  height: 2rem;
  width: 2rem;
}

.c7 .material-symbols-sharp,
.c7 .material-symbols-rounded {
  color: inherit;
}

.c7:hover:not(:disabled) {
  background-color: #EEEEEE;
}

.c7:active:not(:disabled) {
  background-color: #E2E2E2;
}

.c7:disabled {
  opacity: 0.5;
  cursor: initial;
}

.c2 {
  font-size: 0.875rem;
  line-height: 1.375rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  word-break: break-word;
  color: #000000;
  font-family: roboto;
}

.c4 {
  font-size: 0.875rem;
  line-height: 1.375rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  word-break: break-word;
  color: #000000;
  font-family: roboto;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c6 {
  background-color: transparent;
  border: none;
  cursor: default;
}

.c0 {
  height: 2.75rem;
  display: flex;
  background-color: #F6F6F6;
  border-radius: 4px;
  overflow: hidden;
}

.c1 {
  display: flex;
  align-items: center;
  background-color: #EEEEEE;
  padding: 0 1rem;
  flex-shrink: 0;
}

.c3 {
  display: flex;
  align-items: center;
  padding: 0 1rem;
  overflow: hidden;
}

.c5 {
  display: flex;
  align-items: center;
  padding-right: 0.5rem;
  margin-left: auto;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <p
        class="c2"
        font-family="roboto"
      >
        title
      </p>
    </div>
    <div
      class="c3"
    >
      <p
        class="c4"
        font-family="roboto"
      >
        description
      </p>
    </div>
    <div
      class="c5"
    >
      <div
        class="c6"
        data-state="closed"
      >
        <button
          class="c7"
          variant="neutral"
        >
          <div
            class="c8"
            data-testid="icon-container"
          >
            <span
              class="c9 material-symbols-rounded"
              data-testid="icon"
            >
              content_copy
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
